### PR TITLE
Enable coverage reporting for test runners

### PR DESCRIPTION
**Pull request by m-Peter**
Work towards https://github.com/onflow/developer-grants/issues/132

**Complementary PRs:**
- https://github.com/onflow/flow-cli/pull/827
- https://github.com/onflow/cadence/pull/2296

## Description

- Enable coverage reporting for test runners
- Include also the test script name, when printing test results for each script.
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence-lint/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: test/test_runner.go
Comment (turbolent) [inline test/test_runner.go:503]:
```suggestion
	fmt.Fprintf(&sb, "Test results: %q\n", scriptPath)
```

Comment (m-Peter) [reply test/test_runner.go]: Thanks :pray: Fixed in https://github.com/onflow/cadence-tools/pull/78/commits/b0f0774a5b3365c40d206b226f73beedf69b21e5

File: test/test_framework_test.go
Comment (turbolent) [inline test/test_framework_test.go:2617]: Maybe only return this for `StringLocation("./FooContract.cdc")`

**Review by turbolent — APPROVED** [review]
Nice!

**Comment by turbolent** [comment]
Depends on onflow/cadence#2296, so need to wait for a release that includes it

File: test/test_runner.go
Comment (SupunS) [inline test/test_runner.go:238]:
```suggestion
	config := runtime.Config{
		CoverageReportingEnabled: r.coverageReport != nil,
	}
```

**Review by SupunS — APPROVED** [review]
👌

File: test/test_framework_test.go
Comment (m-Peter) [reply test/test_framework_test.go]: Good catch :+1:
Fixed in https://github.com/onflow/cadence-tools/pull/78/commits/a7caee505d22fcc903ce4c0e752287e5df5244ce

File: test/test_runner.go
Comment (m-Peter) [reply test/test_runner.go]: Good catch :+1: 
Fixed in https://github.com/onflow/cadence-tools/pull/78/commits/408b2904c6eb5cc287389c797121d0160c2af0c6

**Comment by sideninja** [comment]
@m-Peter this CI fails, probably need to update the Cadence.

**Comment by m-Peter** [comment]
@sideninja Yes, I think the release flow is probably this:

1. Release a new `cadence` version (to include coverage features)
2. Release a new `cadence-tools` version (has a dependency on `cadence`)
3. Release a new `flow-cli` version (has a dependency on both `cadence` & `cadence-tools`

**Comment by sideninja** [comment]
@turbolent 
> Release a new cadence version (to include coverage features)

Do you have any ETA on that?

File: test/test_runner.go
Comment (m-Peter) [inline test/test_runner.go:246]: We will probably need to exclude this location also

```go
r.coverageReport.ExcludeLocation(stdlib.TestContractLocation)
```
to avoid "noise" for integration tests that import the built-in `Test` contract.

**Comment by m-Peter** [comment]
I was also thinking of randomizing test execution [here](https://github.com/onflow/cadence-tools/blob/master/test/test_runner.go#L158), with something like:

```go
functions := make(
    []*ast.FunctionDeclaration,
    len(program.Program.FunctionDeclarations()),
)
copy(functions, program.Program.FunctionDeclarations())
rand.Shuffle(len(functions), func(i, j int) {
    functions[i], functions[j] = functions[j], functions[i]
})

for _, funcDecl := range functions {
    ...
}
```

What are your thoughts on this?

**Comment by sideninja** [comment]
> I was also thinking of randomizing test execution [here](https://github.com/onflow/cadence-tools/blob/master/test/test_runner.go#L158), with something like:
> 
> ```go
> functions := make(
>     []*ast.FunctionDeclaration,
>     len(program.Program.FunctionDeclarations()),
> )
> copy(functions, program.Program.FunctionDeclarations())
> rand.Shuffle(len(functions), func(i, j int) {
>     functions[i], functions[j] = functions[j], functions[i]
> })
> 
> for _, funcDecl := range functions {
>     ...
> }
> ```
> 
> What are your thoughts on this?

Hmm I'm not sure how I feel about tests not being deterministic, I can see how you could expose flakiness but at the same time it will be hard to identify the source of it I guess due to different execution.

**Comment by turbolent** [comment]
@sideninja The plan is to cut a Cadence release today/early this week

**Comment by m-Peter** [comment]
The code coverage seems to be working quite well, with the unit testing functionality provided by the Cadence testing framework.
However, when I used the integration testing functionality, I realized that deployed contracts are "hidden" behind a proxy (`I.Test`). I have created a [gist](https://gist.github.com/m-Peter/d139fce78c5b1c28153910f580d5de18) to showcase that.

After that, I was toying around with the integration testing functionality, and the `Test` contract. By taking most of the tests from [test_framework_test.go](https://github.com/onflow/cadence-tools/blob/master/test/test_framework_test.go), I was able to calculate its [code coverage](https://gist.github.com/m-Peter/c16d6f0ead4a6b3495635a479626b672) :sweat_smile: We could probably utilize this somehow. It was fun anyway, to learn its API.

File: test/test_runner.go
Comment (m-Peter) [reply test/test_runner.go:248]: Sounds good! Fixed in https://github.com/onflow/cadence-tools/pull/78/commits/4706ae1b9227f2ec5839f8be402370e122624f77

**Comment by m-Peter** [comment]
> > I was also thinking of randomizing test execution [here](https://github.com/onflow/cadence-tools/blob/master/test/test_runner.go#L158), with something like:
> > ```go
> > functions := make(
> >     []*ast.FunctionDeclaration,
> >     len(program.Program.FunctionDeclarations()),
> > )
> > copy(functions, program.Program.FunctionDeclarations())
> > rand.Shuffle(len(functions), func(i, j int) {
> >     functions[i], functions[j] = functions[j], functions[i]
> > })
> > 
> > for _, funcDecl := range functions {
> >     ...
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What are your thoughts on this?
> 
> Hmm I'm not sure how I feel about tests not being deterministic, I can see how you could expose flakiness but at the same time it will be hard to identify the source of it I guess due to different execution.

That's fair, it can be obtrusive if it is done automatically. However, in a stateful paradigm, I think it can uncover some edge cases that are hard to manifest otherwise. If the test suite is executed in a randomized order, the random seed will be printed out, and the developer can then provide this seed, in order to reproduce the failures.

**Comment by sideninja** [comment]
> That's fair, it can be obtrusive if it is done automatically. However, in a stateful paradigm, I think it can uncover some edge cases that are hard to manifest otherwise. If the test suite is executed in a randomized order, the random seed will be printed out, and the developer can then provide this seed, in order to reproduce the failures.

Yeah, it has an upside to indicate issues, but to my experience the wish for that is usually left to a developer to either state in the tests as part of the testing framework to randomize the execution (which in this case I don't think it will be possible since I don't believe there's an API for it) or by providing an execution flag, which might be applicable here. I will leave the decision to you, just wanted to raise the concern I saw.

**Comment by m-Peter** [comment]
> Yeah, it has an upside to indicate issues, but to my experience the wish for that is usually left to a developer to either state in the tests as part of the testing framework to randomize the execution (which in this case I don't think it will be possible since I don't believe there's an API for it) or by providing an execution flag, which might be applicable here. I will leave the decision to you, just wanted to raise the concern I saw.

Makes sense :+1: I was also thinking about an execution flag as well, to allow developers to specify a randomized execution order. Let's wait and see what Bastian and Supun think about it also.

**Comment by sideninja** [comment]
Was #2296 already released? is this still blocked by that?

**Comment by m-Peter** [comment]
@sideninja It was first included in this release: https://github.com/onflow/cadence/releases/tag/v0.34.0 . Let me see if I can it to work with that version.

**Comment by m-Peter** [comment]
@sideninja Bumping only the `cadence` version, doesn't seem to work. We have to also wait for a release of `flow-go` that includes `cadence` >= `v0.34.0`. I see on master branch that it was updated to `v0.36.0` (https://github.com/onflow/flow-go/commit/4cd7028bcc6481b53f2e44557897905a3e2a5694). At least, that's my understanding :sweat_smile:

**Comment by turbolent** [comment]
Yes, flow-go has been updated. The Emulator was just updated in https://github.com/onflow/flow-emulator/pull/330. 

The Emulator will need a couple more improvements (see PR feedback), then after it has been released, I think @dsainati1 will continue updating the downstream dependencies, including the tools here.

**Comment by turbolent** [comment]
#86 should unblock this PR

**Comment by m-Peter** [comment]
I will give it a shot! Thanks for the heads up :raised_hands:

**Comment by m-Peter** [comment]
Well done @turbolent :clap: Rebasing on master, after https://github.com/onflow/cadence-tools/pull/86 was merged, seems to have made all checks successful.

**Pull request merged**
